File: src/main/java/kr/jongyeol/jsBot/LogBuilder.java
package kr.jongyeol.jsBot;

import mx.kenzie.eris.api.entity.*;
import mx.kenzie.eris.api.entity.message.Button;

import java.time.Instant;
import java.util.ArrayList;

import static kr.jongyeol.jsBot.DiscordBot.bot;

public class LogBuilder {
    private static final String nullMessage = "null";

    public static LogBuilder newError(User user) {
        return new LogBuilder(user, "오류가 발생하였습니다.", true);
    }

    public static LogBuilder newError() {
        return newError(bot.getAPI().getSelf());
    }

    private boolean inline = true;

    private final boolean error;
    private final Embed embed;

    public LogBuilder(User user, String message) {
        this(user, message, false);
    }

    public LogBuilder(String message) {
        this(bot.getAPI().getSelf(), message, false);
    }

    private LogBuilder(User user, String message, boolean error) {
        this.error = error;
        embed = new Embed();
        embed.author.name = Utility.getUserName(user);
        embed.author.icon_url = Utility.getAvatar(user);
        embed.color = error ? 0xFF0000 : 0x00FF00;
        embed.description = message;
        embed.footer.text = embed.author.name;
        embed.footer.icon_url = embed.author.icon_url;
        embed.timestamp = Instant.now().toString();
    }

    public LogBuilder setColor(int color) {
        embed.color = color;
        return this;
    }

    public LogBuilder setDefaultInline(boolean inline) {
        this.inline = inline;
        return this;
    }

    public LogBuilder addField(String name, String value) {
        return addField(name, value, inline);
    }

    public LogBuilder addField(String name, String value, boolean inline) {
        if(value == null) value = nullMessage;
        embed.addField(new Embed.Field(name, value, inline));
        return this;
    }

    public LogBuilder addField(String name, Object object) {
        return addField(name, object, inline);
    }

    public LogBuilder addField(String name, Object object, boolean inline) {
        return addField(name, object == null ? nullMessage : object.toString(), inline);
    }

    public LogBuilder addField(String name, Channel channel) {
        return addField(name, channel, inline);
    }

    public LogBuilder addField(String name, Channel channel, boolean inline) {
        if(channel.name == null) channel.await(60000);
        return addField(name, channel.mention() + "(" + channel.name + ")", inline);
    }

    public LogBuilder addField(String name, User user) {
        return addField(name, user, inline);
    }

    public LogBuilder addField(String name, User user, boolean inline) {
        return addField(name, user.mention() + "(" + Utility.getUserName(user) + ")", inline);
    }

    public LogBuilder addField(String name, Member member) {
        return addField(name, member.user, inline);
    }

    public LogBuilder addField(String name, Member member, boolean inline) {
        return addField(name, member.user, inline);
    }

    public LogBuilder addField(String name, int value) {
        return addField(name, value, inline);
    }

    public LogBuilder addField(String name, int value, boolean inline) {
        return addField(name, value + "", inline);
    }

    public LogBuilder addField(String name, mx.kenzie.eris.api.entity.Message message) {
        return addField(name, message, inline);
    }

    public LogBuilder addField(String name, Message message, boolean inline) {
        Channel channel = bot.getAPI().getChannel(message.channel_id);
        channel.await();
        return addField(name, "https://discord.com/channels/" + channel.guild_id + "/" +
            message.channel_id + "/" + message.id, inline);
    }

    public LogBuilder addField(String name, Button button) {
        return addField(name, button, inline);
    }

    public LogBuilder addField(String name, Button button, boolean inline) {
        return addField(name, button.label + "(" + button.custom_id + ")", inline);
    }

    public LogBuilder addField(String name, boolean value) {
        return addField(name, value, inline);
    }

    public LogBuilder addField(String name, boolean value, boolean inline) {
        return addField(name, value ? "true" : "false", inline);
    }

    public LogBuilder addField(String name, Role role) {
        return addField(name, role, inline);
    }

    public LogBuilder addField(String name, Role role, boolean inline) {
        return addField(name, role.mention() + "(" + role.name + ")", inline);
    }

    public LogBuilder addField(Exception e) {
        ArrayList<StringBuilder> builders = new ArrayList<>();
        StringBuilder trace = new StringBuilder();
        builders.add(trace);
        for(StackTraceElement element : e.getStackTrace()) {
            trace.append("\tat ").append(element).append("\n");
            if(trace.length() > 900) {
                trace = new StringBuilder();
                builders.add(trace);
            }
        }
        Embed.Field[] fields = new Embed.Field[builders.size() + 2];
        int i = 0;
        fields[i++] = new Embed.Field("Exception Type", e.getClass().getName());
        fields[i++] = new Embed.Field("Exception Message", e.getMessage());
        for(StringBuilder tracer : builders) fields[i++] = new Embed.Field("Exception Stack Trace", tracer.toString());
        embed.addField(fields);
        return this;
    }

    public LogBuilder addBlankField() {
        return addBlankField(false);
    }

    public LogBuilder addBlankField(boolean inline) {
        embed.addField(new Embed.Field("", "", inline));
        return this;
    }

    public void send() {
        bot.getAPI().sendMessage(JSettings.getInstance().getLogChannel(), new Message(error ? "<@447333460382842880>" : null, embed));
    }
}

File: JsBot/LogBuilder.cs
using NetCord;
using NetCord.Rest;

namespace JsBot;

public class LogBuilder {
	private const string NullMessage = "null";

	public static LogBuilder NewError(User user) => new(user, "오류가 발생하였습니다.", true);
	public static LogBuilder NewError() => NewError(DiscordBot.Client.Cache.User!);

	private bool _inline = true;
	private readonly bool _error;
	private readonly EmbedProperties _embed;

	public LogBuilder(User user, string message) : this(user, message, false) {
	}

	public LogBuilder(string message) : this(DiscordBot.Client.Cache.User!, message, false) {
	}

	private LogBuilder(User user, string message, bool error) {
		_error = error;
		string name = Utility.GetUserName(user);
		string avatar = Utility.GetAvatar(user);
		_embed = new EmbedProperties()
			.WithAuthor(new EmbedAuthorProperties().WithName(name).WithIconUrl(avatar))
			.WithColor(new Color(error ? 0xFF0000 : 0x00FF00))
			.WithDescription(message)
			.WithFooter(new EmbedFooterProperties().WithText(name).WithIconUrl(avatar))
			.WithTimestamp(DateTimeOffset.UtcNow);
	}

	public LogBuilder SetDefaultInline(bool inline) {
		_inline = inline;
		return this;
	}

	public LogBuilder AddField(string name, string? value) => AddField(name, value, _inline);

	public LogBuilder AddField(string name, string? value, bool inline) {
		_embed.AddFields(new EmbedFieldProperties().WithName(name).WithValue(value ?? NullMessage).WithInline(inline));
		return this;
	}

	public LogBuilder AddField(string name, object? value) => AddField(name, value, _inline);
	public LogBuilder AddField(string name, object? value, bool inline) => AddField(name, value?.ToString(), inline);

	public LogBuilder AddField(string name, bool value) => AddField(name, value, _inline);
	public LogBuilder AddField(string name, bool value, bool inline) => AddField(name, value ? "true" : "false", inline);

	public LogBuilder AddField(string name, int value) => AddField(name, value, _inline);
	public LogBuilder AddField(string name, int value, bool inline) => AddField(name, value.ToString(), inline);

	public LogBuilder AddField(string name, ulong value) => AddField(name, value, _inline);
	public LogBuilder AddField(string name, ulong value, bool inline) => AddField(name, value.ToString(), inline);

	public LogBuilder AddField(string name, User user) => AddField(name, user, _inline);
	public LogBuilder AddField(string name, User user, bool inline) => AddField(name, Utility.GetUserMention(user.Id) + "(" + Utility.GetUserName(user) + ")", inline);

	public LogBuilder AddField(string name, TextChannel channel) => AddField(name, channel, _inline);
	public LogBuilder AddField(string name, TextChannel channel, bool inline) => AddField(name, Utility.GetChannelMention(channel.Id), inline);

	public LogBuilder AddField(string name, RestMessage message) => AddField(name, message, _inline);
	public LogBuilder AddField(string name, RestMessage message, bool inline) =>
		AddField(name, "https://discord.com/channels/" + JSettings.Instance.GuildId + "/" + message.ChannelId + "/" + message.Id, inline);

	public LogBuilder AddField(string name, ButtonProperties button) => AddField(name, button, _inline);
	public LogBuilder AddField(string name, ButtonProperties button, bool inline) => AddField(name, button.Label + "(" + button.CustomId + ")", inline);

	public LogBuilder AddField(string name, Role role) => AddField(name, role, _inline);
	public LogBuilder AddField(string name, Role role, bool inline) => AddField(name, Utility.GetRoleMention(role.Id) + "(" + role.Name + ")", inline);

	public LogBuilder AddField(Exception e) {
		List<string> traces = [];
		System.Text.StringBuilder trace = new();
		traces.Add("");
		foreach(string line in (e.StackTrace ?? "").Split('\n')) {
			if(trace.Length + line.Length > 1024) {
				traces[^1] = trace.ToString();
				trace.Clear();
				traces.Add("");
			}
			trace.Append(line).Append('\n');
		}
		traces[^1] = trace.ToString();
		_embed.AddFields(new EmbedFieldProperties().WithName("Exception Type").WithValue(e.GetType().FullName ?? e.GetType().Name).WithInline(_inline));
		_embed.AddFields(new EmbedFieldProperties().WithName("Exception Message").WithValue(e.Message).WithInline(_inline));
		foreach(string tracer in traces) _embed.AddFields(new EmbedFieldProperties().WithName("Exception Stack Trace").WithValue(tracer.Length == 0 ? NullMessage : tracer).WithInline(_inline));
		return this;
	}

	public LogBuilder AddBlankField() => AddBlankField(false);

	public LogBuilder AddBlankField(bool inline) {
		_embed.AddFields(new EmbedFieldProperties().WithName("").WithValue("").WithInline(inline));
		return this;
	}

	public void Send() {
		MessageProperties message = new MessageProperties().WithEmbeds([_embed]);
		if(_error) message.Content = "<@447333460382842880>";
		_ = DiscordBot.Rest.SendMessageAsync(JSettings.Instance.LogChannel, message);
	}
}